test/functorch/test_control_flow.py

```diff
@@ -259,6 +262,1063 @@ def false_fn(x):
+        c = torch.randn(4, requires_grad=True)
+
+        for pred, fn in zip(
+            [torch.tensor(False), torch.tensor(True)], [false_fn, true_fn]
+        ):
+            with self.assertRaisesRegex(
+                torch._dynamo.exc.UncapturedHigherOrderOpError,
+                "Cond doesn't work unless it is captured completely with torch.compile.",
+            ):
+                cond(pred, true_fn, false_fn, ({"t": [a, {"b": b}, (c,)]},))
+
+    def test_cond_autograd_same_pytree_output(self):
+        # TODO: If inside the dictionary, inside the list, the first element
+        # is composed of the multiplication, then the requires_grad attribute is
+        # set to False and thus the tracing of the cond errors out.
+        def true_fn(x):
+            # return {"res": [x['t'][0] * x['t'][1]['b'], (x['t'][2][0],)]}
+            return {"res": [x["t"][0], (x["t"][2][0],)]}
+
+        def false_fn(x):
+            return {"res": [x["t"][1]["b"], (x["t"][2][0],)]}
+
+        a = torch.randn(4, requires_grad=True)
+        b = torch.randn(4, requires_grad=True)
+        c = torch.randn(4, requires_grad=True)
+
+        for pred, fn in zip(
+            [torch.tensor(False), torch.tensor(True)], [false_fn, true_fn]
+        ):
+            result = cond(pred, true_fn, false_fn, ({"t": [a, {"b": b}, (c,)]},))
+            result_exp = fn({"t": [a, {"b": b}, (c,)]})
+            self.assertEqual(result, result_exp)
+
+            result_flat, _ = pytree.tree_flatten(result)
+            result_exp_flat, _ = pytree.tree_flatten(result_exp)
+
+            grad_out = [torch.ones_like(g) for g in result_flat]
+            expected_grads = torch.autograd.grad(result_exp_flat, (c,), grad_out)
+            grads = torch.autograd.grad(result_flat, (c,), grad_out)
+            self.assertEqual(expected_grads, grads)
+
+        def f(pred):
+            result = cond(pred, true_fn, false_fn, ({"t": [a, {"b": b}, (c,)]},))
+            return result
+
+        gm = make_fx(f, tracing_mode="symbolic", _allow_non_fake_inputs=True)(pred)
+        self.assertExpectedInline(
+            gm.code.strip(),
+            """\
+def forward(self, pred_1):
+    true_graph_0 = self.true_graph_0
+    false_graph_0 = self.false_graph_0
+    _tensor_constant0 = self._tensor_constant0
+    _tensor_constant1 = self._tensor_constant1
+    _tensor_constant2 = self._tensor_constant2
+    cond = torch.ops.higher_order.cond(pred_1, true_graph_0, false_graph_0, (_tensor_constant0, _tensor_constant1, _tensor_constant2));  pred_1 = true_graph_0 = false_graph_0 = _tensor_constant0 = _tensor_constant1 = _tensor_constant2 = None
+    getitem = cond[0]
+    getitem_1 = cond[1];  cond = None
+    view = torch.ops.aten.view.default(getitem, [4]);  getitem = None
+    view_1 = torch.ops.aten.view.default(getitem_1, [4]);  getitem_1 = None
+    return {'res': [view, (view_1,)]}""",  # noqa: B950
+        )
+
+    def test_cond_autograd_torch_nn_module(self):
```

Comment:
This testcase still shows detach operations for certain operations, for example torch.nn.functional.relu. Is this something that we need to worry about?

Reply:
Yeah, detach is expected behavior. This happens when "autograd saves a Tensor for backwards". What tensor to be saved varies from each operators to another.